Data/Map/Strict.hs

```diff
@@ -15,19 +15,73 @@
 -- Maintainer  :  libraries@haskell.org
 -- Portability :  portable
 --
--- An efficient implementation of ordered maps from keys to values
--- (dictionaries).
 --
--- API of this module is strict in both the keys and the values.
--- If you need value-lazy maps, use "Data.Map.Lazy" instead.
--- The 'Map' type is shared between the lazy and strict modules,
--- meaning that the same 'Map' value can be passed to functions in
--- both modules (although that is rarely needed).
+-- = Strict Maps (aka dictionaries)
+--
+-- The @'Map' k v@ type represents an /ordered/ map indexed by keys of type @k@
+-- containing /strict/ values of type @v@.
+--
+-- When deciding if this is the correct data structure to use, consider:
+--
+-- * The API of this module is strict in both the keys and the values. If you
+-- need /value-lazy/ maps, use "Data.Map.Lazy" instead.
+--
+-- * If your key type is @Int@, use "Data.IntMap.Strict" instead.
+--
+-- * If you don't care about ordering, use @Data.HashMap.Strict@ from the
+-- <https://hackage.haskell.org/package/unordered-containers unordered-containers>
+-- package instead.
+--
+-- This module is intended to be imported qualified, to avoid name clashes with
+-- Prelude functions:
+--
+-- > import qualified Data.Map.Strict as Map
+--
+-- Note that the implementation is /left-biased/, the elements of a first
+-- argument are always preferred to the second, for example in 'union' or
+-- 'insert'.
+--
+--
+-- == Detailed performance information
+--
+-- The amortized running time is given for each operation, with /n/ referring to
+-- the number of entries in the map.
+--
+-- Benchmarks comparing "Data.Map.Strict" with other dictionary implementations
+-- can be found at https://github.com/haskell-perf/dictionaries.
+--
+--
+-- == Strictness propertios
+--
+-- This module satisfies the following strictness properties:
+--
+-- 1. Key arguments are evaluated to WHNF;
+--
+-- 2. Keys and values are evaluated to WHNF before they are stored in
+--    the map.
```

Comment:
Remove point 1 and all examples attempting to illustrate it. Instead, write something like

    -- As described, 'Map's are strict in their keys. In addition, functions in this module evaluate
    -- values before storing them in maps. More precisely:
    --
    --     If all values in all 'Map' arguments to a function are in WHNF, then
    --     all values in all 'Map' results will be in WHNF.

We need to be a bit careful because some operations (such as `union`) will reuse subtrees without touching them.